src/basic/namespace-util.c

```diff
@@ -210,14 +213,42 @@ int userns_acquire(const char *uid_map, const char *gid_map) {
 
         /* Forks off a process in a new userns, configures the specified uidmap/gidmap, acquires an fd to it,
          * and then kills the process again. This way we have a userns fd that is not bound to any
-         * process. We can use that for file system mounts and similar. */
+         * process. We can use that for file system mounts and similar.
+         * If the owner of the new namespace is specified, setuid to it in the child. Use a pipe to sync and
+         * ensure setuid() happens first, then unshare(), then the mapping update. */
 
-        r = safe_fork("(sd-mkuserns)", FORK_CLOSE_ALL_FDS|FORK_DEATHSIG|FORK_NEW_USERNS, &pid);
+        if (pipe2(pipe_fd, 0) < 0)
+                return log_error_errno(errno, "Failed to open bidirectional pipe: %m");
+
+        /* If we have to set a different owner we have to setuid before unsharing the new namespace, otherwise
+         * the kernel will refuse the operation, so don't fork directly in the new user namespace in that case.
+         * If we don't have to change the owner then it easier and we can let the clone set up the namespace
+         * instead via the apposite flag. */
+        if (owner == UID_INVALID)
+                flags |= FORK_NEW_USERNS;
+
+        r = safe_fork_full("(sd-mkuserns)", (int[]) { pipe_fd[1] }, 1, flags, &pid);
         if (r < 0)
                 return r;
-        if (r == 0)
-                /* Child. We do nothing here, just freeze until somebody kills us. */
+        if (r == 0) {
+                /* Child. Signal after setuid() and then just freeze until somebody kills us. */
+                if (owner != UID_INVALID) {
+                        (void) setuid(owner);
+                        (void) unshare(CLONE_NEWUSER);
+                }
+                pipe_fd[1] = safe_close(pipe_fd[1]);
                 freeze();
+        }
+
+        pipe_fd[1] = safe_close(pipe_fd[1]);
+
+        if (owner != UID_INVALID) {
+                r = fd_wait_for_event(pipe_fd[0], POLLHUP, USEC_PER_SEC * 10);
+                if (r < 0)
+                        return r;
+                if (r == 0)
+                        return -ETIMEDOUT;
+        }
 
         xsprintf(path, "/proc/" PID_FMT "/uid_map", pid);
```

Comment:
if the setuid() failed for some reason (RLIMIT_NPROC for example) then you might end up writing into the uid_map of the originating userns. that looks very wrong.

plesae alway check for errors.

Reply:
The originating ns is the root one, so I don't think that can actually happen? But added anyway

Reply:
could happen in an nspawn container that has userns applied, no?